File: Packages/TeamTasksKit/Sources/TeamTasksCore/Services/PlatformPorts.swift
import Foundation

// Platform abstractions implemented by the iOS app (UserNotifications, UserDefaults…)
// and by simple fakes in tests, so all the logic stays testable on Linux.

public enum NotificationAuthorization: Sendable, Hashable {
    case notDetermined
    case denied
    case authorized
}

/// A local notification request.
public struct LocalNotification: Sendable, Hashable, Identifiable {
    /// Stable identifier. Prefixes: `due-` (due-date reminders), `assigned-` (new assignment), `summary-`, and (v2)
    /// `recap-` (the weekly recap).
    public var id: String
    public var title: String
    public var body: String
    /// Delivery date; nil means "deliver now". With `repeatsWeekly`, the first delivery (for a scheduler that does
    /// not know repeating triggers).
    public var fireDate: Date?
    /// Routing info for taps: keys `taskId`, `groupId` (UUID strings).
    public var userInfo: [String: String]
    public var threadId: String?
    /// v2: delivered every week at this time, in the device's calendar and time zone, until removed. The platform
    /// schedules a repeating calendar trigger and ignores `fireDate`.
    public var repeatsWeekly: WeeklyRepeat?

    public init(
        id: String,
        title: String,
        body: String,
        fireDate: Date? = nil,
        userInfo: [String: String] = [:],
        threadId: String? = nil,
        repeatsWeekly: WeeklyRepeat? = nil
    ) {
        self.id = id
        self.title = title
        self.body = body
        self.fireDate = fireDate
        self.userInfo = userInfo
        self.threadId = threadId
        self.repeatsWeekly = repeatsWeekly
    }
}

/// v2: a weekly time of a repeating local notification (the weekly recap, docs/CONTRACTS-V2.md §8): every `weekday`
/// at `hour`:`minute`, in the device's calendar and time zone.
public struct WeeklyRepeat: Sendable, Hashable, Codable {
    /// ISO weekday: 1 = Monday … 7 = Sunday.
    public var weekday: Int
    public var hour: Int
    public var minute: Int

    public init(weekday: Int, hour: Int, minute: Int = 0) {
        self.weekday = weekday
        self.hour = hour
        self.minute = minute
    }

    /// The same day as a Gregorian `DateComponents.weekday`: 1 = Sunday, 2 = Monday … 7 = Saturday.
    public var gregorianWeekday: Int { weekday % 7 + 1 }

    /// The components the platform trigger matches (`UNCalendarNotificationTrigger(dateMatching:repeats: true)`).
    public var dateComponents: DateComponents {
        DateComponents(hour: hour, minute: minute, weekday: gregorianWeekday)
    }

    /// The first time strictly after `date`, in `calendar`'s time zone.
    public func nextDate(after date: Date, calendar: Calendar) -> Date {
        let startOfToday = calendar.startOfDay(for: date)
        let today = calendar.component(.weekday, from: startOfToday)
        let daysAhead = (gregorianWeekday - today + 7) % 7
        for extraWeeks in 0...2 {
            guard let day = calendar.date(byAdding: .day, value: daysAhead + 7 * extraWeeks, to: startOfToday),
                  let candidate = calendar.date(bySettingHour: hour, minute: minute, second: 0, of: day)
            else { continue }
            if candidate > date { return candidate }
        }
        return date.addingTimeInterval(7 * 86_400)
    }
}

public protocol NotificationScheduler: Sendable {
    func authorizationStatus() async -> NotificationAuthorization
    /// Returns true if granted.
    func requestAuthorization() async -> Bool
    /// Identifiers of pending (not yet delivered) requests starting with `prefix`.
    func pendingIdentifiers(prefix: String) async -> [String]
    func add(_ notification: LocalNotification) async throws
    func removePending(ids: [String]) async
    func setBadge(_ count: Int) async
}

/// Small persistent key-value storage (UserDefaults on iOS).
public protocol KeyValueStore: Sendable {
    func data(forKey key: String) -> Data?
    func set(_ data: Data?, forKey key: String)
}

extension KeyValueStore {
    public func value<T: Decodable>(_ type: T.Type, forKey key: String) -> T? {
        guard let data = data(forKey: key) else { return nil }
        return try? JSONDecoder().decode(T.self, from: data)
    }

    public func setValue<T: Encodable>(_ value: T?, forKey key: String) {
        guard let value else {
            set(nil, forKey: key)
            return
        }
        set(try? JSONEncoder().encode(value), forKey: key)
    }
}

/// In-memory `KeyValueStore` (tests, previews, UI tests).
public final class InMemoryKeyValueStore: KeyValueStore, @unchecked Sendable {
    private let lock = NSLock()
    private var storage: [String: Data] = [:]

    public init() {}

    public func data(forKey key: String) -> Data? {
        lock.withLock { storage[key] }
    }

    public func set(_ data: Data?, forKey key: String) {
        lock.withLock { storage[key] = data }
    }
}

/// Current-date provider (injectable for tests). Not named `Clock` to avoid clashing with Swift's `Clock`.
public typealias NowProvider = @Sendable () -> Date

/// Platform services injected into view models.
public struct PlatformServices: Sendable {
    public var notifications: any NotificationScheduler
    public var store: any KeyValueStore
    public var now: NowProvider
    public var calendar: Calendar

    public init(
        notifications: any NotificationScheduler,
        store: any KeyValueStore,
        now: @escaping NowProvider = { Date() },
        calendar: Calendar = .current
    ) {
        self.notifications = notifications
        self.store = store
        self.now = now
        self.calendar = calendar
    }
}

File: android/core/src/main/kotlin/io/github/notkanaa/equipe/core/PlatformPorts.kt
package io.github.notkanaa.equipe.core

import kotlinx.serialization.SerializationException
import kotlinx.serialization.json.Json
import kotlinx.serialization.serializer
import java.time.DayOfWeek
import java.time.Instant
import java.time.LocalDate
import java.time.LocalTime
import java.time.ZoneId
import java.time.ZonedDateTime
import java.time.temporal.WeekFields
import java.util.Locale

// Port of TeamTasksCore/Services/PlatformPorts.swift: platform abstractions implemented by the Android app
// (NotificationManager/WorkManager, DataStore…) and by simple fakes in tests, so all the logic stays testable on the JVM.

enum class NotificationAuthorization {
    NOT_DETERMINED,
    DENIED,
    AUTHORIZED,
}

/** A local notification request. */
data class LocalNotification(
    /** Stable identifier. Prefixes: `due-` (due-date reminders), `assigned-` (new assignment), `summary-`. */
    val id: String,
    val title: String,
    val body: String,
    /** Delivery date; null means "deliver now". */
    val fireDate: Instant? = null,
    /** Routing info for taps: keys `taskId`, `groupId` (UUID strings). */
    val userInfo: Map<String, String> = emptyMap(),
    val threadId: String? = null,
)

interface NotificationScheduler {
    suspend fun authorizationStatus(): NotificationAuthorization

    /** Returns true if granted. */
    suspend fun requestAuthorization(): Boolean

    /** Identifiers of pending (not yet delivered) requests starting with [prefix]. */
    suspend fun pendingIdentifiers(prefix: String): List<String>

    suspend fun add(notification: LocalNotification)

    suspend fun removePending(ids: List<String>)

    suspend fun setBadge(count: Int)
}

/** Small persistent key-value storage (UserDefaults on iOS; DataStore/SharedPreferences on Android). */
interface KeyValueStore {
    /** The bytes stored under [key], or null. */
    fun data(key: String): ByteArray?

    /** Stores [data] under [key]; null removes the key. */
    fun set(key: String, data: ByteArray?)
}

/** JSON codec of [value] / [setValue] (unknown keys ignored, like Swift's `JSONDecoder`). */
val KeyValueStoreJson: Json = Json { ignoreUnknownKeys = true }

/** The JSON value stored under [key] (kotlinx.serialization), or null when absent or undecodable. */
inline fun <reified T> KeyValueStore.value(key: String, json: Json = KeyValueStoreJson): T? {
    val bytes = data(key) ?: return null
    return try {
        json.decodeFromString(serializer<T>(), bytes.decodeToString())
    } catch (error: IllegalArgumentException) { // SerializationException included
        null
    }
}

/** Stores [value] as JSON under [key]; null (or a value that cannot be encoded) removes the key. */
inline fun <reified T> KeyValueStore.setValue(key: String, value: T?, json: Json = KeyValueStoreJson) {
    if (value == null) {
        set(key, null)
        return
    }
    val encoded = try {
        json.encodeToString(serializer<T>(), value).encodeToByteArray()
    } catch (error: SerializationException) {
        null
    }
    set(key, encoded)
}

/** In-memory [KeyValueStore] (tests, previews, UI tests). Thread-safe; stores copies (value semantics). */
class InMemoryKeyValueStore : KeyValueStore {
    private val storage = HashMap<String, ByteArray>()

    override fun data(key: String): ByteArray? = synchronized(storage) { storage[key]?.copyOf() }

    override fun set(key: String, data: ByteArray?) {
        synchronized(storage) {
            if (data == null) storage.remove(key) else storage[key] = data.copyOf()
        }
    }
}

/** Current-date provider (injectable for tests). */
typealias NowProvider = () -> Instant

/**
 * The app's calendar (Swift `Calendar.frenchGregorian(timeZone:)`): Gregorian (ISO chronology of java.time), weeks
 * starting on Monday with at least 4 days in the first week (ISO 8601), `fr_FR` locale, injected time zone.
 *
 * Only java.time APIs available on Android 26 are used (e.g. no `LocalDate.ofInstant`, which is Java 9).
 */
data class AppCalendar(
    val zone: ZoneId,
    val locale: Locale = FRENCH,
) {
    /** Monday first, minimal days in the first week: 4. */
    val weekFields: WeekFields get() = WeekFields.of(DayOfWeek.MONDAY, 4)

    fun zonedDateTime(instant: Instant): ZonedDateTime = instant.atZone(zone)

    /** The calendar day of [instant] in [zone]. */
    fun localDate(instant: Instant): LocalDate = instant.atZone(zone).toLocalDate()

    /** Start of the day of [instant] in [zone] (DST-aware: not always midnight). */
    fun startOfDay(instant: Instant): Instant = startOfDay(localDate(instant))

    fun startOfDay(date: LocalDate): Instant = date.atStartOfDay(zone).toInstant()

    /** The instant of a wall-clock [time] on [date] in [zone] (gaps resolved forward, overlaps to the earlier offset). */
    fun instant(date: LocalDate, time: LocalTime): Instant = ZonedDateTime.of(date, time, zone).toInstant()

    fun isSameDay(lhs: Instant, rhs: Instant): Boolean = localDate(lhs) == localDate(rhs)

    companion object {
        /** `fr_FR`. */
        val FRENCH: Locale = Locale.FRANCE

        val PARIS: ZoneId = ZoneId.of("Europe/Paris")

        /** French Gregorian rules in [zone] (Swift `Calendar.frenchGregorian(timeZone:)`). */
        fun frenchGregorian(zone: ZoneId): AppCalendar = AppCalendar(zone, FRENCH)

        /** French Gregorian rules in the device's current time zone. */
        fun systemDefault(): AppCalendar = frenchGregorian(ZoneId.systemDefault())
    }
}

/** Platform services injected into view models. */
data class PlatformServices(
    val notifications: NotificationScheduler,
    val store: KeyValueStore,
    val now: NowProvider = { Instant.now() },
    val calendar: AppCalendar = AppCalendar.systemDefault(),
)
